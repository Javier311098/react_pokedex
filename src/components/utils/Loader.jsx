import React from "react";
import { CgPokemon } from "react-icons/cg";
export const Loader = () => {
  return (
    <div className="spinner-img-container">
      <CgPokemon className="img-loader" />
      <h2>Loading...</h2>
    </div>
  );
};
