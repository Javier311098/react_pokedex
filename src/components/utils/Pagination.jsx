import React, { useEffect } from "react";
import { AiFillStepBackward, AiFillStepForward } from "react-icons/ai";
import { useDispatch } from "react-redux";
import { useSelector } from "react-redux";
import { currentPage } from "../../redux/actions/pagination";

export const Pagination = () => {
  const {
    pagination: { totalPages, actualPage },
    pokemons: { loading },
  } = useSelector((state) => state);
  const dispatch = useDispatch();

  useEffect(() => {
    const page = JSON.parse(localStorage.getItem("page") || 0);
    dispatch(currentPage(parseInt(page)));
  }, [dispatch]);

  const handleNextPage = () => {
    dispatch(currentPage(Math.min(actualPage + 1, totalPages - 1)));
    localStorage.setItem("page", Math.min(actualPage + 1, totalPages - 1));
  };

  const handleLastPage = () => {
    dispatch(currentPage(Math.max(actualPage - 1, 0)));
    localStorage.setItem("page", Math.max(actualPage - 1, 0));
  };

  return (
    <div className="pagination-container">
      <div className="pagination-container-info">
        <button
          disabled={loading ? true : false}
          className="pagination-buttons"
          onClick={handleLastPage}
        >
          <AiFillStepBackward />
        </button>
        <span>
          {actualPage + 1} of {totalPages}
        </span>
        <button
          disabled={loading ? true : false}
          className="pagination-buttons"
          onClick={handleNextPage}
        >
          <AiFillStepForward />
        </button>
      </div>
    </div>
  );
};
