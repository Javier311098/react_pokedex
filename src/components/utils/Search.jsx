import React, { useState } from "react";
import { BsSearch } from "react-icons/bs";
import { useNavigate } from "react-router-dom";
import Swal from "sweetalert2";

export const Search = () => {
  const [inputValue, setInputValue] = useState("");
  const navigate = useNavigate();

  const handleInputChange = ({ target }) => {
    if (target.value.length >= 15) {
      setInputValue("");
    } else {
      setInputValue(target.value);
    }
  };

  const handleSearch = () => {
    if (inputValue.length >= 1) {
      navigate(`/${process.env.REACT_APP_REPO_NAME}/${inputValue}`);
    } else {
      Swal.fire(
        "Error",
        "You need write a pokemon name or pokedex number",
        "error"
      );
    }
  };

  return (
    <div className="search-container">
      <div className="search-bar">
        <input
          autoComplete="false"
          type="text"
          onChange={handleInputChange}
          onKeyPress={(event) => event.key === "Enter" && handleSearch()}
          value={inputValue}
          name="inputValue"
          placeholder="Buscar pokemon..."
        />
        <button onClick={handleSearch} className="search-button">
          <BsSearch />
        </button>
      </div>
    </div>
  );
};
