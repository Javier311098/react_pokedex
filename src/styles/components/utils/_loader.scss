.spinner-img-container {
  display: flex;
  height: 100vh;
  width: 100vw;
  align-items: center;
  justify-content: center;
  flex-direction: column;
}

.img-loader {
  width: 150px;
  height: 150px;
  animation: spin 1s ease infinite;
}

@keyframes spin {
  0% {
    transform: rotate(0deg);
  }
  100% {
    transform: rotate(360deg);
  }
}
