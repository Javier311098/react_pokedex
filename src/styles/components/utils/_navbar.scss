.navbar-container {
  border-radius: 5px;
  border: 2px solid black;
  height: 120px;
  padding: 5px;
  margin: 5px;
  display: flex;
  align-items: center;
  justify-content: space-evenly;
  flex-direction: column;
}

.navbar-title {
  display: flex;
  align-items: center;
  img {
    width: 50px;
    height: 50px;
  }
}
