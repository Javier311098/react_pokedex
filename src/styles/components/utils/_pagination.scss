.pagination-container {
  display: flex;
  justify-content: center;
}
.pagination-container-info {
  display: flex;
  justify-content: center;
  align-items: center;
  font-size: 20px;
  span {
    font-family: "Pokemon Solid";
    color: $logo-blue;
  }
}

.pagination-buttons {
  background-color: transparent;
  border: none;
  cursor: pointer;
  width: 50px;
  font-size: 25px;
  color: $logo-yellow;

  &:hover {
    color: darken($color: $logo-yellow, $amount: 10);
  }
}
