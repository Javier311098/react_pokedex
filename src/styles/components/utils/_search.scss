.search-container {
  display: flex;
  justify-content: center;
}

.search-bar {
  display: flex;
  justify-content: center;
  input {
    font-size: 20px;
    height: 35px;
    border: none;
    border-radius: 5px;
  }
}

.search-button {
  background-color: transparent;
  color: $logo-blue;
  border-radius: 5px;
  cursor: pointer;
  border: none;
  &:hover {
    background-color: darken($color: white, $amount: 10);
    border: none;
  }
}
