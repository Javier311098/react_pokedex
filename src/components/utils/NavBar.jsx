import React from "react";
import { Pagination } from "./Pagination";
import { Search } from "./Search";

export const NavBar = () => {
  return (
    <div className="navbar-container animate__animated animate__fadeInDown">
      <div className="navbar-title">
        <h1>Pokedex</h1>
        <img
          alt="pokedex"
          src="http://i.ebayimg.com/images/g/HCgAAOxy63FSrBOH/s-l500.jpg"
        />
      </div>
      <Pagination />
      <Search />
    </div>
  );
};
