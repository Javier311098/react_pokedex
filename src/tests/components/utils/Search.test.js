/* eslint-disable testing-library/no-debugging-utils */

import { shallow } from "enzyme";
import Swal from "sweetalert2";
import { Search } from "../../../components/utils/Search";

const mockNavigate = jest.fn();
jest.mock("react-router-dom", () => ({
  ...jest.requireActual("react-router-dom"),
  useNavigate: () => mockNavigate,
}));

jest.mock("sweetalert2", () => ({
  fire: jest.fn(),
}));

describe("Testing Search ||Component", () => {
  test("should render the component", () => {
    const wrapper = shallow(<Search />);
    expect(wrapper).toMatchSnapshot();
    expect(wrapper.find(".search-container").exists()).toBe(true);
  });

  test("should get an error alert when input value its empty", () => {
    const wrapper = shallow(<Search />);
    wrapper.find("input").simulate("keypress", { key: "Enter" });
    expect(Swal.fire).toHaveBeenCalledWith(
      "Error",
      "You need write a pokemon name or pokedex number",
      "error"
    );
  });

  test("should change inputValue to pokemon name", () => {
    const wrapper = shallow(<Search />);
    const pokemon = "pikachu";
    wrapper.find("input[name='inputValue']").simulate("change", {
      target: {
        name: "inputValue",
        value: pokemon,
      },
    });
    expect(wrapper.find("input").prop("value")).toBe(pokemon);
  });

  test("should change inputValue to empty when write more than 15 characters", () => {
    const wrapper = shallow(<Search />);
    wrapper.find("input[name='inputValue']").simulate("change", {
      target: {
        name: "inputValue",
        value: "kbjehorhioioriprjpfrw",
      },
    });
    expect(wrapper.find("input").prop("value")).toBe("");
  });

  test("should change the page when pokemon value its found", () => {
    const pokemon = "pikachu";
    const wrapper = shallow(<Search />);
    wrapper.find("input[name='inputValue']").simulate("change", {
      target: {
        name: "inputValue",
        value: pokemon,
      },
    });
    wrapper.find("button").prop("onClick")();
    expect(mockNavigate).toHaveBeenCalledWith(
      `/${process.env.REACT_APP_REPO_NAME}/${pokemon}`
    );
  });
});
