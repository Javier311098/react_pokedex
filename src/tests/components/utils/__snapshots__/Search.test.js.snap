// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing Search ||Component should render the component 1`] = `
<div
  className="search-container"
>
  <div
    className="search-bar"
  >
    <input
      autoComplete="false"
      name="inputValue"
      onChange={[Function]}
      onKeyPress={[Function]}
      placeholder="Buscar pokemon..."
      type="text"
      value=""
    />
    <button
      className="search-button"
      onClick={[Function]}
    >
      <BsSearch />
    </button>
  </div>
</div>
`;
